### PR TITLE
Add HDFS reading

### LINKED ISSUE #18199: Feature to read csv from hdfs:// URL
When running pandas in AWS, The following works perfectly fine:
```python
pd.read_csv("s3://mybucket/data.csv")
```
But running the following, does not:
```python
pd.read_csv("hdfs:///tmp/data.csv")
```

It would be a good user experience to allow for the hdfs:// schema too similar to how http, ftp, s3, and file are valid schemas right now.

### PR DESCRIPTION
**Pull request by AbdealiLoKo**
- Add HDFS reading using `libhdfs3`

- [x] closes #18199
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/hdfs.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+""" s3 support for remote file interactivity """
+from pandas import compat
+try:
+    import hdfs3
+except:
+    raise ImportError("The hdfs3 library is required to handle hdfs files")
+
+if compat.PY3:
```
Comment (jreback) [inline pandas/io/hdfs.py]: these should be moved to ``pandas/compat/__init__.py`` (and same change in pandas.io.s3fs)

File: pandas/io/parquet.py [hunk 2]
```diff
@@ -25,6 +25,10 @@ def get_engine(engine):
         except ImportError:
             pass
 
+        raise ImportError("unable to find a usable engine\n"
+                          "tried using: pyarrow, fastparquet\n\n"
```
Comment (jreback) [inline pandas/io/parquet.py]: why are you changing this? IIRC this is a separeate issue, make it a separate PR

File: pandas/io/hdfs.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+""" s3 support for remote file interactivity """
+from pandas import compat
+try:
+    import hdfs3
```
Comment (jreback) [inline pandas/io/hdfs.py]: you need to add hdfs3 on *some* builds (you can put on the same as s3fs) is fine in pandas/ci

File: pandas/io/common.py [hunk 4]
```diff
@@ -314,11 +328,17 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
...
-        need_text_wrapping = (BytesIO, S3File)
+        need_text_wrapping.append(S3File)
+    except ImportError:
+        pass
+    try:
+        from hdfs3 import HDFile
+        need_text_wrapping.append(HDFile)
     except ImportError:
-        need_text_wrapping = (BytesIO,)
+        pass
```
Comment (jreback) [inline pandas/io/common.py]: you need

``need_text_wrapping = tuple(need_text_wrapping)`` before ``handles``

File: doc/source/whatsnew/v0.22.0.txt [hunk 5]
```diff
@@ -155,6 +155,7 @@ I/O
 ^^^
 
 - :func:`read_html` now rewinds seekable IO objects after parse failure, before attempting to parse with a new parser. If a parser errors and the object is non-seekable, an informative error is raised suggesting the use of a different parser (:issue:`17975`)
+- :func:`read_csv` now supports reading from hdfs by giving "hdfs:///tmp/data.csv". The hadoop configs will try to be automatically found. The configs can also be mentioned using the format "hdfs://nodenamehost:nodenameport/tmp/data.csv"
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: add in the Other Enhancements section. Add documentionat in io.rst (where s3 docs are found)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
@AbdealiJK I cherry-picked your first commit (as this was an easy change) and merged that in a separate PR (https://github.com/pandas-dev/pandas/pull/18717).

Can you update the hdfs part according to the comments?

**Comment by AbdealiJK** [comment]
Thanks Jeff. Modifying this is on my ToDo - a bit pre-occupied at the
moment.  Will do it soon.

**Comment by AbdealiJK** [comment]
For the unit test: I need some help in setting up HDFS inside Travis.
The way other orgs are doing this is by using their own dockers: [Example of hdfs3's CI](https://github.com/dask/hdfs3/blob/master/continuous_integration/Dockerfile)

Pandas does not seem to have it's own docker ... so a little confused about this.

Manually I've tested the following:
 - If hdfs3 is not installed, Throws:
```
ImportError: The hdfs3 library is required to handle hdfs files
```
 - If hdfs3 is installed but libhdfs3 is not installed, Throws:
```
ImportError: Can not find the shared library: libhdfs3.so
See installation instructions at http://hdfs3.readthedocs.io/en/latest/install.html
```
 - If hdfs3 is installed it works for the code: pd.read_csv("hdfs://localhost:9000/tmp/a.csv")
 - If hdfs3 is installed it works if export HADOOP_CONF_DIR=/usr/local/Cellar/hadoop/2.7.0/libexec/etc/hadoop/ is done and pd.read_csv("hdfs:///tmp/a.csv") is provided

**Comment by jreback** [comment]
if you can rebase. this would require a testing setup similar to this:

https://github.com/dask/dask/tree/master/continuous_integration/hdfs

**Comment by jreback** [comment]
closing as stale. nice idea, but we have to be able to automatically test this.

**Pull request closed without merging**